lib/components/user/favorite-location.js

```diff
@@ -0,0 +1,236 @@
+import React, { Component } from 'react'
+import {
+  Button,
+  DropdownButton,
+  FormGroup,
+  Glyphicon,
+  InputGroup,
+  MenuItem
+} from 'react-bootstrap'
+import FontAwesome from 'react-fontawesome'
+import styled from 'styled-components'
+
+import FavoriteLocationField from '../form/favorite-location-field'
+
+// Styles
+const StyledDropdown = styled(DropdownButton)`
+  background-color: #eee;
+  width: 40px;
+`
+const NewLocationDropdownButton = styled(StyledDropdown)`
+  background-color: #337ab7;
+  color: #fff;
+`
+const StyledAddon = styled(InputGroup.Addon)`
+  width: 40px;
+`
+
+// Helper filter functions.
+export const isHome = loc => loc.type === 'home'
+export const isWork = loc => loc.type === 'work'
+
+// Defaults for home and work
+export const BLANK_HOME = {
+  address: '',
+  icon: 'home',
+  name: '',
+  type: 'home'
+}
+export const BLANK_WORK = {
+  address: '',
+  icon: 'briefcase',
+  name: '',
+  type: 'work'
+}
+
+/**
+ * Creates a new favorite location object from a location returned by LocationField, with the specified type and icon.
+ */
+function makeFavoriteLocation (baseLocation, type, icon) {
+  const { lat, lon, name } = baseLocation
+  return {
+    address: name,
+    icon,
+    lat,
+    lon,
+    name,
+    type
+  }
+}
+
+/**
+ * Renders a user favorite location,
+ * and lets the user edit the details (address, type, TODO: nickname) of the location.
+ */
+class FavoriteLocation extends Component {
+  constructor () {
+    super()
+    this.state = {
+      newLocation: null
+    }
+  }
+
+  _handleDelete = () => {
+    const { arrayHelpers, index, isFixed, location } = this.props
+    if (isFixed) {
+      // For 'Home' and 'Work', replace with the default blank locations instead of deleting.
+      let newLocation
+      if (isHome(location)) {
+        newLocation = BLANK_HOME
+      } else if (isWork(location)) {
+        newLocation = BLANK_WORK
+      }
+
+      this._handleLocationChange({ location: newLocation })
+    } else {
+      arrayHelpers.remove(index)
+    }
+  }
+
+  _handleLocationChange = ({ location }) => {
+    const { arrayHelpers, index, location: oldLocation } = this.props
+    const { icon, type } = oldLocation
+    arrayHelpers.replace(index, makeFavoriteLocation(location, type, icon))
```

Comment:
The updated UI mocks for user account show that we're moving away from making changes and then clicking a 'Save Preferences' button. I don't think we should invest more in this current approach. I'd like to see these handlers make PUT calls to update the user account with the list of locations. I'm also wondering if it might be good to have `UserLocation.class` extend Model so that we can get ids on these objects (separate concern from what's at hand).

Reply:
I think we will need an overall discussion when we commision the work to remove the 'Save Preferences' button.

Reply:
This current approach (not saving anything until 'Save prefs' is clicked) is not working well from a UX perspective, so it's really not beneficial to build this component around that behavior.

Reply:
I agree on the principle. I think it is best to revise that event handler around the time we make the change to the backend.